db/db_impl/db_impl_write.cc

```diff
@@ -893,12 +891,13 @@ void DBImpl::MemTableInsertStatusCheck(const Status& status) {
 
 Status DBImpl::PreprocessWrite(const WriteOptions& write_options,
                                bool* need_log_sync,
+                               log::Writer** back_writer,
                                WriteContext* write_context) {
-  mutex_.AssertHeld();
   assert(write_context != nullptr && need_log_sync != nullptr);
   Status status;
 
   if (error_handler_.IsDBStopped()) {
+    InstrumentedMutexLock l(&mutex_);
```

Comment:
All the checks in `PreprocessWrite` (e.g `error_handler_.IsDBStopped()` here) are supposed to guarded by `mutex_` previously. You need to make each of them lock-free.

Reply:
Sorry, I misread the change. `IsDBStopped` and some of the other checks are changed to be lock-free. I'll double check.